README.md
# Bottlerocket ECS Updater

The Bottlerocket ECS Updater is a service you can install into your ECS cluster that helps you keep your Bottlerocket container instances up to date.
When installed, the Bottlerocket ECS Updater will periodically query each Bottlerocket container instance to find whether an update is available and drain tasks while an update is in progress.
Updates to Bottlerocket are rolled out in [waves](https://github.com/bottlerocket-os/bottlerocket/tree/develop/sources/updater/waves) to reduce the impact of issues; the container instances in your cluster may not all see updates at the same time.

## Installation

You can install the Bottlerocket ECS Updater into your cluster with the provided [CloudFormation template](stacks/bottlerocket-ecs-updater.yaml).
The following information is required when creating the CloudFormation stack:

* The name of the ECS cluster where you are running Bottlerocket container instances
* The name of the CloudWatch Logs log group where the Bottlerocket ECS Updater will send its logs
* At least one subnet ID that has Internet access (which does not need to be shared with the rest of your cluster)

When installed, the CloudFormation template will create the following resources in your account:

* A task definition for the Bottlerocket ECS Updater
* A CloudWatch Events scheduled rule to execute the Bottlerocket ECS Updater
* An IAM role for the Bottlerocket ECS Updater task itself as well as roles for Fargate and CloudWatch Events
* SSM documents to query and execute updates on Bottlerocket instances

## Getting Started

To install the Bottlerocket ECS Updater, you will need to fetch some information first.

### Subnet info

You should either have a default virtual private cloud (VPC) or have already
[created a VPC](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/get-set-up-for-amazon-ecs.html#create-a-vpc)
in your account.

To find your default VPC, run this command.
(If you use an AWS region other than "us-west-2", make sure to change that.)

```sh
aws ec2 describe-vpcs \
   --region us-west-2 \
   --filters=Name=isDefault,Values=true \
   | jq --raw-output '.Vpcs[].VpcId'
```

If you want to use a different VPC you created, run this to get the ID for your VPC.
Make sure to change VPC_NAME to the name of the VPC you created.
(If you use an EC2 region other than "us-west-2", make sure to change that too.)

```sh
aws ec2 describe-vpcs \
   --region us-west-2 \
   --filters=Name=tag:Name,Values=VPC_NAME \
   | jq --raw-output '.Vpcs[].VpcId'
```

Next, run this to get information about the subnets in your VPC.
It will give you a list of the subnets and tell you whether each is public or private.
Make sure to change VPC_ID to the value you received from the previous command.
(If you use an EC2 region other than "us-west-2", make sure to change that too.)

```sh
aws ec2 describe-subnets \
   --region us-west-2 \
   --filter=Name=vpc-id,Values=VPC_ID \
   | jq '.Subnets[] | {id: .SubnetId, public: .MapPublicIpOnLaunch, az: .AvailabilityZone}'
```

You'll want to pick at least one and save it for the launch command later.
Make sure the subnets you select have Internet access so the updater can reach its dependencies.
Public subnets usually have Internet access via an [Internet gateway](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html) while private subnets may be configured with NAT.
For more information, see [the VPC user guide](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html#vpc-igw-internet-access).

We recommend picking several subnets in different availability zones.
However, if you want to launch in a specific availability zone, make sure you pick a subnet that matches; the AZ is listed right below the public/private status.

### Log Group

You can either choose an existing log group or create a new one to get your ECS updater logs.

You can run this to get the list of existing log-groups
```sh
aws logs describe-log-groups
```

You'll want to pick one and save it for the installation command later.

If you want to create a new log group, run this (Make sure to provide LOG_GROUP_NAME)
```sh
aws logs create-log-group --log-group-name LOG_GROUP_NAME
```

### Install

Now we can install the [CloudFormation template](stacks/bottlerocket-ecs-updater.yaml) to start the ECS updater for your cluster!

There are a few values to make sure you change in this command:
* CLUSTER_NAME: the name of the cluster you want ECS updater to manage Bottlerocket instances in
* SUBNET_IDS: a comma-separated list of the subnets you selected earlier
* LOG_GROUP_NAME: the log group name you selected or created earlier

```sh
aws cloudformation deploy \
    --stack-name "bottlerocket-ecs-updater" \
    --template-file "./stacks/bottlerocket-ecs-updater.yaml" \
    --capabilities CAPABILITY_NAMED_IAM \
    --parameter-overrides \
    ClusterName="CLUSTER_NAME" \
    Subnets="SUBNET_IDS" \
    LogGroupName="LOG_GROUP_NAME"
```

## How it works

The Bottlerocket ECS Updater is designed to run as a scheduled Fargate task that queries, drains, and performs updates in your ECS cluster.
A rule in CloudWatch Events periodically launches the updater as a new Fargate task.
The updater queries the ECS API to discover all the container instances in your cluster and filters for Bottlerocket instances by reading the `bottlerocket.variant` attribute.
For each Bottlerocket instance found, the updater executes an SSM document that queries for available updates using the `apiclient update check` command.
When an update is available, the updater checks to see whether the tasks currently running on the container instance are part of a [service](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html) and eligible for replacement.
If all the tasks are part of a service, the updater marks the container instance for [draining](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-instance-draining.html) and waits for the tasks to be successfully drained.
After the container instance has been drained, the updater executes an SSM document to download the update, apply the update, and reboot.
Finally, the updater will mark the container instance as active and move on to the next one.

## Troubleshooting

When installed with the provided CloudFormation template, the logs for the updater will be available the CloudWatch Logs group you configured.
Checking the logs is a good first step in understanding why something happened or didn't happen.

### Why do only some of my Bottlerocket instances have an update available?

Updates to Bottlerocket are rolled out in [waves](https://github.com/bottlerocket-os/bottlerocket/tree/develop/sources/updater/waves) to reduce the impact of issues; the container instances in your cluster may not all see updates at the same time.
You can check whether an update is available on your instance by running the `apiclient update check` command from within the [control](https://github.com/bottlerocket-os/bottlerocket#control-container) or [admin](https://github.com/bottlerocket-os/bottlerocket#admin-container) container.

### My Bottlerocket instance has an update available.  Why didn't the Bottlerocket ECS Updater update it?

The Bottlerocket ECS Updater attempts to update container instances without disrupting the workloads in your cluster.
Applying an update to Bottlerocket requires a reboot.
To avoid disruption in your cluster, the Bottlerocket ECS Updater uses the [container instance draining](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-instance-draining.html) feature of ECS.
A container instance may be skipped for update when:

* _A non-service task is running._
  Non-service tasks are not automatically replaced when they are stopped.
  To avoid disrupting a critical workload, the Bottlerocket ECS Updater will not stop a non-service task.
* _No spare capacity is available in the cluster._
  The service scheduler attempts to replace the tasks according to the service's deployment configuration parameters, `minimumHealthyPercent` and `maximumPercent`.
  If stopping a task would reduce the running count below your service's `minimumHealthyPercent`, ECS will not stop the task.
  The Bottlerocket ECS Updater will wait for draining to complete for a fixed period of time (currently 25 minutes).
  If draining has not completed by the end of the period, the updater will restore the instance and move to the next one.
* _Draining takes too long._
  The Bottlerocket ECS Updater will wait for draining to complete for a fixed period of time (currently 25 minutes).
  If draining has not completed by the end of the period, the updater will restore the instance and move to the next one.
  The time it takes for a task to be stopped is related to the `stopTimeout` task definition parameter and to any associated resources like load balancers.
  If your tasks are taking too long to drain, you can ensure that your task responds to `SIGTERM`, shorten the `stopTimeout`, or shorten the load balancer's health check and deregistration delay settings.
* _Bottlerocket version is too old._
  The Bottlerocket ECS Updater uses newer [`apiclient update` commands](https://github.com/bottlerocket-os/bottlerocket#update-api) that were added in version [1.0.5](https://github.com/bottlerocket-os/bottlerocket/blob/develop/CHANGELOG.md#v105-2021-01-15).
  The SSM commands will fail if your Bottlerocket OS version is less than 1.0.5.
  Instances running Bottlerocket versions less than 1.0.5 need to be manually updated.
* _Too many instances are in the cluster._
  The Bottlerocket ECS Updater currently supports clusters of up to 50 container instances.
  If the updater is configured to target a cluster with more than 50 instances, some instances may not be updated.

### Why do new container instances launch with older Bottlerocket versions?

The Bottlerocket ECS Updater performs in-place updates for instances in your ECS cluster.
The updater does not influence how those instances are launched.
If you use an auto-scaling group to launch your instances, you can update the AMI ID in your launch configuration or launch template to use a newer version of Bottlerocket.

Note: We do not recommend using the Bottlerocket ECS Updater in conjunction with EC2 Spot.
The ECS Updater is designed to keep services safe from interruption by updating one instance at a time.
With the short average lifetime of Spot instances, the updater may not update them until relatively late in their life, meaning they may not be up to date when serving your application.

## Security

See [CONTRIBUTING](CONTRIBUTING.md#security-issue-notifications) for more information.

## License

This project is dual licensed under either the Apache-2.0 License or the MIT license, your choice.



CHANGELOG.md
# 0.2.2

* Bump version of Go to 1.19.2
* Upgrade `github.com/aws/aws-sdk-go` dependency to `v1.44.137`
* Upgrade `github.com/stretchr/testify` dependency to `v1.8.1`

# 0.2.1

* Bump version of Go to 1.19.1 and upgrade dependencies

# 0.2.0

* Add support for clusters larger than 50 container instances.
* Add after-action summary and done message to log output.
* Add check to reduce the chance of concurrent runs.

Note: In the Bottlerocket ECS updater v0.1.0 release, support for clusters was limited to 50 container instances. In this release, clusters larger than 50 container instances are now supported. :tada: 

# 0.1.0

Initial release of the **Bottlerocket ECS updater** - A service to automatically manage Bottlerocket updates in an Amazon ECS cluster.

The Bottlerocket ECS updater is designed to help you safely automate the routine maintenance of updating the Bottlerocket instances in your cluster.
The updater's safety features include:

* Only tasks that are part of a [service](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html) will be interrupted.
  Container instances with non-service tasks are skipped for upgrade so no critical workloads will be automatically interrupted.
* Only container instances in the [ACTIVE state](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-instance-draining.html) will be upgrade.
  Instances that have been placed into the DRAINING state are skipped for upgrade so other maintenance or debugging can be performed without interruption.

In this first release of the updater, the following considerations should be kept in mind:

* Only clusters of up to 50 container instances are supported.
  If the updater is configured to target a cluster with more than 50 instances, some instances may not be updated.
* When configuring the provided CloudFormation template, ensure that the CloudWatch log group already exists.
  The updater will not automatically create the log group and a missing log group will cause the updater to fail to run.
  When creating a log group, you can configure your desired log retention settings.

See the [README](README.md) for additional information.

updater/aws.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/request"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/aws/aws-sdk-go/service/ecs"
	"github.com/aws/aws-sdk-go/service/ssm"
)

const (
	ecsPageSize          = 100
	ssmPageSize          = 50
	updateStateIdle      = "Idle"
	updateStateStaged    = "Staged"
	updateStateAvailable = "Available"
	updateStateReady     = "Ready"
	waiterDelay          = time.Duration(15) * time.Second
	waiterMaxAttempts    = 100
	// If this time is reached and the ssm command has not already started running, it will not run.
	deliveryTimeoutSeconds = 600
)

type instance struct {
	instanceID          string
	containerInstanceID string
	bottlerocketVersion string
}

type checkOutput struct {
	UpdateState     string `json:"update_state"`
	ActivePartition struct {
		Image struct {
			Version string `json:"version"`
		} `json:"image"`
	} `json:"active_partition"`
}

type ECSAPI interface {
	ListContainerInstancesPages(*ecs.ListContainerInstancesInput, func(*ecs.ListContainerInstancesOutput, bool) bool) error
	DescribeContainerInstances(input *ecs.DescribeContainerInstancesInput) (*ecs.DescribeContainerInstancesOutput, error)
	UpdateContainerInstancesState(input *ecs.UpdateContainerInstancesStateInput) (*ecs.UpdateContainerInstancesStateOutput, error)
	ListTasks(input *ecs.ListTasksInput) (*ecs.ListTasksOutput, error)
	DescribeTasks(input *ecs.DescribeTasksInput) (*ecs.DescribeTasksOutput, error)
	WaitUntilTasksStoppedWithContext(ctx aws.Context, input *ecs.DescribeTasksInput, opts ...request.WaiterOption) error
}

type SSMAPI interface {
	WaitUntilCommandExecutedWithContext(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error
	SendCommand(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error)
	GetCommandInvocation(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error)
}

type EC2API interface {
	WaitUntilInstanceStatusOk(input *ec2.DescribeInstanceStatusInput) error
}

func (u *updater) alreadyRunning(family string) (bool, error) {
	log.Print("Checking for running updater tasks")
	list, err := u.ecs.ListTasks(&ecs.ListTasksInput{
		Cluster: &u.cluster,
		Family:  aws.String(family),
	})
	if err != nil {
		return false, fmt.Errorf("failed to list running updater tasks: %w", err)
	}
	if len(list.TaskArns) > 1 {
		return true, nil
	}
	log.Println("This is the only running updater.")
	return false, nil
}

func (u *updater) listContainerInstances() ([]*string, error) {
	log.Printf("Listing active container instances in cluster %q", u.cluster)
	containerInstances := make([]*string, 0)
	input := &ecs.ListContainerInstancesInput{
		Cluster: &u.cluster,
		Status:  aws.String(ecs.ContainerInstanceStatusActive),
	}
	if err := u.ecs.ListContainerInstancesPages(input, func(output *ecs.ListContainerInstancesOutput, lastpage bool) bool {
		containerInstances = append(containerInstances, output.ContainerInstanceArns...)
		return true
	}); err != nil {
		return nil, fmt.Errorf("failed to list container instances: %w", err)
	}
	log.Printf("Found %d container instances in the cluster", len(containerInstances))
	return containerInstances, nil
}

// filterBottlerocketInstances filters container instances and returns list of
// instances that are running Bottlerocket OS
func (u *updater) filterBottlerocketInstances(instances []*string) ([]instance, error) {
	log.Printf("Filtering container instances running Bottlerocket OS")
	bottlerocketInstances := make([]instance, 0)
	errCount := 0
	var lastErr error
	pageCount, err := eachPage(len(instances), ecsPageSize, func(start, stop int) error {
		resp, err := u.ecs.DescribeContainerInstances(&ecs.DescribeContainerInstancesInput{
			Cluster:            &u.cluster,
			ContainerInstances: instances[start:stop],
		})
		// count errors per page.
		if err != nil {
			log.Printf("Failed to describe container instances from %d to %d: %v", start, stop, err)
			errCount++
			lastErr = err
			return nil
		}
		for _, containerInstance := range resp.ContainerInstances {
			if containsAttribute(containerInstance.Attributes, "bottlerocket.variant") {
				bottlerocketInstances = append(bottlerocketInstances, instance{
					instanceID:          aws.StringValue(containerInstance.Ec2InstanceId),
					containerInstanceID: aws.StringValue(containerInstance.ContainerInstanceArn),
				})
				log.Printf("Bottlerocket instance %q detected.", aws.StringValue(containerInstance.Ec2InstanceId))
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	// check if every page had an error; errors are only fatal if each page failed.
	if errCount == pageCount {
		return nil, fmt.Errorf("failed to describe any container instances: %w", lastErr)
	}
	return bottlerocketInstances, nil
}

// containsAttribute checks if a slice of ECS Attributes struct contains a specified name.
func containsAttribute(attrs []*ecs.Attribute, searchString string) bool {
	for _, attr := range attrs {
		if aws.StringValue(attr.Name) == searchString {
			return true
		}
	}
	return false
}

// eachPage defines batch processing boundaries for handling paginated results of API calls.
func eachPage(inputLen int, size int, fn func(start, stop int) error) (int, error) {
	pageCount := 0
	for start := 0; start < inputLen; start += size {
		stop := start + size
		if stop > inputLen {
			stop = inputLen
		}
		if err := fn(start, stop); err != nil {
			return 0, err
		}
		pageCount++
	}
	return pageCount, nil
}

// filterAvailableUpdates returns a list of instances that have updates available
func (u *updater) filterAvailableUpdates(bottlerocketInstances []instance) ([]instance, error) {
	log.Printf("Filtering instances with available updates")
	// make slice of Bottlerocket instances to use with SendCommand and checkCommandOutput
	instances := make([]string, 0)
	for _, inst := range bottlerocketInstances {
		instances = append(instances, inst.instanceID)
	}

	var lastErr error
	errCount := 0
	candidates := make([]instance, 0)
	pageCount, err := eachPage(len(instances), ssmPageSize, func(start, stop int) error {
		commandID, err := u.sendCommand(instances[start:stop], u.checkDocument)
		if err != nil {
			// errors here are considered non-fatal.
			log.Printf("Failed to send document %s: %v", u.checkDocument, err)
			errCount++
			lastErr = err
			return nil
		}
		for _, inst := range bottlerocketInstances[start:stop] {
			commandOutput, err := u.getCommandResult(commandID, inst.instanceID)
			if err != nil {
				// errors here are considered non-fatal
				log.Printf("Failed to get output for command %s, document %s and instance %q: %v", commandID, u.checkDocument, inst, err)
				continue
			}
			output, err := parseCommandOutput(commandOutput)
			if err != nil {
				log.Printf("Failed to parse command output %q for instance %q: %v", string(commandOutput), inst, err)
				continue
			}
			if output.UpdateState == updateStateAvailable || output.UpdateState == updateStateReady {
				inst.bottlerocketVersion = output.ActivePartition.Image.Version
				candidates = append(candidates, inst)
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	if errCount == pageCount {
		return nil, fmt.Errorf("all attempts to send SSM document %s failed: %w", u.checkDocument, lastErr)
	}
	return candidates, nil
}

// eligible checks the eligibility of container instance for update. It's eligible
// if all the running tasks were started by a service.
func (u *updater) eligible(containerInstance string) (bool, error) {
	log.Printf("Checking eligiblity for update of container instance %q", containerInstance)
	list, err := u.ecs.ListTasks(&ecs.ListTasksInput{
		Cluster:           &u.cluster,
		ContainerInstance: aws.String(containerInstance),
	})
	if err != nil {
		return false, fmt.Errorf("failed to list tasks: %w", err)
	}
	taskARNs := list.TaskArns
	if len(list.TaskArns) == 0 {
		return true, nil
	}

	desc, err := u.ecs.DescribeTasks(&ecs.DescribeTasksInput{
		Cluster: &u.cluster,
		Tasks:   taskARNs,
	})
	if err != nil {
		return false, fmt.Errorf("failed to describe tasks: %w", err)
	}
	for _, listResult := range desc.Tasks {
		startedBy := aws.StringValue(listResult.StartedBy)
		if !strings.HasPrefix(startedBy, "ecs-svc/") {
			log.Printf("Container instance %q has a non-service task running: %s", containerInstance, aws.StringValue(listResult.TaskArn))
			return false, nil
		}
	}
	return true, nil
}

func (u *updater) drainInstance(containerInstance string) error {
	log.Printf("Starting drain on container instance %q", containerInstance)
	resp, err := u.ecs.UpdateContainerInstancesState(&ecs.UpdateContainerInstancesStateInput{
		Cluster:            &u.cluster,
		ContainerInstances: aws.StringSlice([]string{containerInstance}),
		Status:             aws.String("DRAINING"),
	})
	if err != nil {
		return fmt.Errorf("failed to change instance state to DRAINING: %w", err)
	}
	if len(resp.Failures) != 0 {
		log.Printf("There are API failures in draining the container instance %q, therefore attempting to"+
			" re-activate", containerInstance)
		err = u.activateInstance(containerInstance)
		if err != nil {
			log.Printf("Instance failed to re-activate after failing to change state to DRAINING: %v", err)
		}
		return fmt.Errorf("failures in API call: %v", resp.Failures)
	}
	log.Printf("Container instance state changed to DRAINING")

	err = u.waitUntilDrained(containerInstance)
	if err != nil {
		log.Printf("Container instance %q failed to drain, therefore attempting to re-activate", containerInstance)
		err2 := u.activateInstance(containerInstance)
		if err2 != nil {
			log.Printf("Instance failed to re-activate after failing to wait for drain to complete: %v", err2)
		}
		return fmt.Errorf("error while waiting to drain: %w", err)
	}
	log.Printf("Container instance %q drained successfully!", containerInstance)
	return nil
}

func (u *updater) activateInstance(containerInstance string) error {
	resp, err := u.ecs.UpdateContainerInstancesState(&ecs.UpdateContainerInstancesStateInput{
		Cluster:            &u.cluster,
		ContainerInstances: aws.StringSlice([]string{containerInstance}),
		Status:             aws.String("ACTIVE"),
	})
	if err != nil {
		return fmt.Errorf("failed to change state to ACTIVE: %w", err)
	}
	if len(resp.Failures) != 0 {
		if aws.StringValue(resp.Failures[0].Reason) == "INACTIVE" {
			log.Printf("Container instance %q is in INACTIVE state", containerInstance)
			return nil
		}
		return fmt.Errorf("API failures while activating: %v", resp.Failures)
	}
	log.Printf("Container instance %q state changed to ACTIVE successfully!", containerInstance)
	return nil
}

func (u *updater) waitUntilDrained(containerInstance string) error {
	log.Printf("Waiting for container instance %q to drain", containerInstance)
	list, err := u.ecs.ListTasks(&ecs.ListTasksInput{
		Cluster:           &u.cluster,
		ContainerInstance: aws.String(containerInstance),
	})
	if err != nil {
		return fmt.Errorf("failed to list tasks: %w", err)
	}
	taskARNs := list.TaskArns

	if len(taskARNs) == 0 {
		log.Printf("No tasks to drain")
		return nil
	}

	return u.ecs.WaitUntilTasksStoppedWithContext(aws.BackgroundContext(), &ecs.DescribeTasksInput{
		Cluster: &u.cluster,
		Tasks:   taskARNs,
	},
		request.WithWaiterMaxAttempts(waiterMaxAttempts),
		request.WithWaiterDelay(request.ConstantWaiterDelay(waiterDelay)),
	)
}

// updateInstance starts an update process on an instance.
func (u *updater) updateInstance(inst instance) error {
	log.Printf("Starting update on instance %q", inst.instanceID)
	ec2IDs := []string{inst.instanceID}
	log.Printf("Checking current update state of instance %q", inst.instanceID)

	commandID, err := u.sendCommand(ec2IDs, u.checkDocument)
	if err != nil {
		return fmt.Errorf("failed to send check command: %w", err)
	}
	output, err := u.getCommandResult(commandID, inst.instanceID)
	if err != nil {
		return fmt.Errorf("failed to get check command output: %w", err)
	}
	check, err := parseCommandOutput(output)
	if err != nil {
		return fmt.Errorf("failed to parse command output %q: %w", string(output), err)
	}

	switch check.UpdateState {
	case updateStateIdle:
		log.Printf("No new update available for instance %q", inst.instanceID)
		return nil
	case updateStateStaged:
		return fmt.Errorf("unexpected update state %q; skipping instance", check.UpdateState)
	case updateStateAvailable:
		log.Printf("Starting update apply on instance %q", inst.instanceID)
		_, err := u.sendCommand(ec2IDs, u.applyDocument)
		if err != nil {
			return fmt.Errorf("failed to send update apply command: %w", err)
		}
	case updateStateReady:
		log.Printf("Update is previously applied on instance %q", inst.instanceID)
	default:
		return fmt.Errorf("unknown update state %q", check.UpdateState)
	}

	// occasionally instance goes into reboot before reporting command output, therefore
	// we do not poll for command output. Instead we rely on verifyUpdate to confirm update
	// success or failure.
	log.Printf("Sending SSM document %q on instance %q", u.rebootDocument, inst.instanceID)
	// SendCommand is directly called here because we do not want to wait on command complete.
	resp, err := u.ssm.SendCommand(&ssm.SendCommandInput{
		DocumentName:    aws.String(u.rebootDocument),
		DocumentVersion: aws.String("$DEFAULT"),
		InstanceIds:     aws.StringSlice(ec2IDs),
		TimeoutSeconds:  aws.Int64(deliveryTimeoutSeconds),
	})
	if err != nil {
		return fmt.Errorf("failed to send reboot command: %w", err)
	}
	rebootID := *resp.Command.CommandId
	log.Printf("SSM document %q posted with command ID %q", u.rebootDocument, rebootID)

	// added some sleep time for reboot to start before we check instance state
	time.Sleep(15 * time.Second)
	err = u.waitUntilOk(inst.instanceID)
	if err != nil {
		return fmt.Errorf("failed to reach Ok status after reboot: %w", err)
	}
	return nil
}

// verifyUpdate verifies if instance was properly updated
func (u *updater) verifyUpdate(inst instance) (bool, error) {
	log.Println("Verifying update by checking there is no new version available to update" +
		" and validate the active version")
	ec2IDs := []string{inst.instanceID}
	updateStatus, err := u.sendCommand(ec2IDs, u.checkDocument)
	if err != nil {
		return false, fmt.Errorf("failed to send update check command: %w", err)
	}

	updateResult, err := u.getCommandResult(updateStatus, inst.instanceID)
	if err != nil {
		return false, fmt.Errorf("failed to get check command output: %w", err)
	}
	output, err := parseCommandOutput(updateResult)
	if err != nil {
		return false, fmt.Errorf("failed to parse command output %q, manual verification required: %w", string(updateResult), err)
	}
	updatedVersion := output.ActivePartition.Image.Version
	if updatedVersion == inst.bottlerocketVersion {
		log.Printf("Container instance %q did not update, its current "+
			"version %s and updated version %s are the same", inst.containerInstanceID, inst.bottlerocketVersion, updatedVersion)
		return false, nil
	} else if output.UpdateState == updateStateAvailable {
		log.Printf("Container instance %q was updated to version %q successfully, however another newer version was recently released;"+
			" Instance will be updated to newer version in next iteration.", inst.containerInstanceID, updatedVersion)
		return true, nil
	} else {
		log.Printf("Container instance %q updated to version %q", inst.containerInstanceID, updatedVersion)
	}
	return true, nil
}

func (u *updater) sendCommand(instanceIDs []string, ssmDocument string) (string, error) {
	log.Printf("Sending SSM document %q", ssmDocument)
	resp, err := u.ssm.SendCommand(&ssm.SendCommandInput{
		DocumentName:    aws.String(ssmDocument),
		DocumentVersion: aws.String("$DEFAULT"),
		InstanceIds:     aws.StringSlice(instanceIDs),
		TimeoutSeconds:  aws.Int64(deliveryTimeoutSeconds),
	})
	if err != nil {
		return "", fmt.Errorf("send command failed: %w", err)
	}
	commandID := *resp.Command.CommandId
	log.Printf("SSM document %q posted with command id %q", ssmDocument, commandID)

	// Wait for the sent commands to complete.
	wg := sync.WaitGroup{}
	instanceCount := len(instanceIDs)
	errChan := make(chan error, instanceCount)
	for _, v := range instanceIDs {
		log.Printf("Waiting for command %q to complete for instance %q", commandID, v)
		wg.Add(1)
		go func(instanceID string) {
			defer wg.Done()
			err = u.ssm.WaitUntilCommandExecutedWithContext(aws.BackgroundContext(), &ssm.GetCommandInvocationInput{
				CommandId:  aws.String(commandID),
				InstanceId: aws.String(instanceID),
			},
				request.WithWaiterMaxAttempts(waiterMaxAttempts),
				request.WithWaiterDelay(request.ConstantWaiterDelay(waiterDelay)))
			if err != nil {
				errChan <- err
				log.Printf("Error encountered while awaiting document %q execution for instance: %q: %s", ssmDocument, instanceID, err)
				u.logCommmandOutput(commandID, instanceID)
			}
		}(aws.StringValue(&v))
	}
	wg.Wait()
	close(errChan)

	errCount := 0
	for err = range errChan {
		errCount++
		if errCount == instanceCount {
			return "", fmt.Errorf("too many failures while awaiting document execution: %w", err)
		}
	}
	return commandID, nil
}

func (u *updater) getCommandResult(commandID string, instanceID string) ([]byte, error) {
	resp, err := u.ssm.GetCommandInvocation(&ssm.GetCommandInvocationInput{
		CommandId:  aws.String(commandID),
		InstanceId: aws.String(instanceID),
	})
	if err != nil {
		return nil, fmt.Errorf("failed to retrieve command invocation output: %w", err)
	}
	commandResults := []byte(aws.StringValue(resp.StandardOutputContent))
	if aws.StringValue(resp.Status) != ssm.CommandInvocationStatusSuccess {
		return nil, fmt.Errorf("command %s has not reached success status, current status %q", commandID, aws.StringValue(resp.Status))
	}
	return commandResults, nil
}

// logCommmandOutput logs the ssm command invocation response
func (u *updater) logCommmandOutput(commandID string, instanceID string) {
	resp, err := u.ssm.GetCommandInvocation(&ssm.GetCommandInvocationInput{
		CommandId:  aws.String(commandID),
		InstanceId: aws.String(instanceID),
	})
	if err != nil {
		log.Printf("Failed to get invocation output for instance %q: %v", instanceID, err)
	}
	log.Printf("Invocation output for instance %q: %#q", instanceID, resp)
}

// waitUntilOk takes an EC2 ID as a parameter and waits until the specified EC2 instance is in an Ok status.
func (u *updater) waitUntilOk(ec2ID string) error {
	log.Printf("Waiting for instance %q to reach Ok status", ec2ID)
	return u.ec2.WaitUntilInstanceStatusOk(&ec2.DescribeInstanceStatusInput{
		InstanceIds: []*string{aws.String(ec2ID)},
	})
}

// parseCommandOutput takes raw bytes of ssm command output and converts it into a struct
func parseCommandOutput(commandOutput []byte) (checkOutput, error) {
	output := checkOutput{}
	err := json.Unmarshal(commandOutput, &output)
	if err != nil {
		return output, fmt.Errorf("failed to unmarshal json: %w", err)
	}
	if output.UpdateState == "" || output.ActivePartition.Image.Version == "" {
		return output, fmt.Errorf("mandatory fields are not available")
	}
	return output, nil
}
